### PR TITLE
Add AppContext data support for DefaultMaxConnectionsPerServer

### LINKED ISSUE #59809: Add AppContext switch to set default MaxConnectionsPerServer
### Problem description

Larger applications migrating from .NET Framework to .NET Core can hit **port exhaustion** caused by the fact that `MaxConnectionsPerServer` on `HttpClient` defaults to INF on .NET Core (while it is configurable app-wide in .NET Framework via `ServicePointManager`).
The fix is often not straightforward when the app uses multiple in-house and external libraries (incl. NuGet packages and various SDKs). Many/most of the libraries do not expose way to configure their own HttpClient.
As a result, applications tend to use hacks like Reflection to get to the HttpClients created inside the libraries. They typically do that after they hit production issues.

In ideal world, we would provide clear guidance what libraries should do/expose and all libraries would follow it. This will take some non-trivial time and will not address older mature libraries which are not super-actively maintained.

### Suggested solution

Introduce a workaround -- AppContext switch that would allow to se default value of MaxConnectionsPerServer in your application.
While not ideal, it should help address larger apps to get the problems under control in reasonable way.

### PR DESCRIPTION
**Pull request by rzikm**
Fixes #59809.

This PR adds support for configuring default value of `MaxConnectionsPerServer` on `HttpClientHandler` and `WinHttpClientHandler`. 

Usage:

```cs
public class Program
{
    static void Main(string[] args)
    {
        AppContext.SetData("System.Net.Http.HttpHandlerDefaults.DefaultMaxConnectionsPerServer", 1);
        var handler = new HttpClientHandler();
        System.Console.WriteLine(handler.MaxConnectionsPerServer); // prints 1
    }
}
```

The AppContext.SetData function must be called before the first instance of `HttpClientHandler` is created, later updates will not take effect.

### CODE REVIEW

**Comment by rzikm** [comment]
I originally wanted to reuse `AppContextConfigHelper.GetInt32Config` for the implementation, but that code uses internal methods on `bool` that cannot be called from outside `System.Private.CoreLib`. I wasn't sure that splitting the one function I needed to a separate file was the right approach.

**Comment by ManickaP** [comment]
Is there a reason why it cannot be in https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Http/src/System/Net/Http/GlobalHttpSettings.cs where we have the other "AppSwitches"?
cc @antonfirsov

File: src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.cs [hunk 1]
```diff
@@ -2013,5 +2014,19 @@ public async Task SendAsync_InvalidRequestUri_Throws()
...
         }
+
+        [ConditionalTheory(typeof(RemoteExecutor), nameof(RemoteExecutor.IsSupported))]
+        [InlineData(true)]
+        [InlineData(false)]
+        public void AppContextSetData_SetDefaultMaxConnectionsPerServer(bool asInt)
+        {
+            RemoteExecutor.Invoke((asInt) =>
+            {
+                object data = asInt == "true" ? 1 : "1";
```
Comment (campersau) [inline src/libraries/Common/tests/System/Net/Http/HttpClientHandlerTest.cs:2025]: `true.ToString() == "True"`
```suggestion
                object data = asInt == "True" ? 1 : "1";
```

or

```suggestion
                object data = asInt == Boolean.TrueString ? 1 : "1";
```

**Comment by rzikm** [comment]
> Is there a reason why it cannot be in https://github.com/dotnet/runtime/blob/main/src/libraries/System.Net.Http/src/System/Net/Http/GlobalHttpSettings.cs where we have the other "AppSwitches"? cc @antonfirsov Anton Firszov FTE

Yes, I didn't know this existed :D I think I will move the default to an envvar so that it is more consistent with the other settings we have there.

**Review by antonfirsov — APPROVED** [review]
Looks good with the latest changes.

File: src/libraries/System.Net.Http/src/System/Net/Http/GlobalHttpSettings.cs [hunk 2]
```diff
@@ -77,5 +77,25 @@ private static double GetHttp2StreamWindowScaleThresholdMultiplier()
...
 #endif
+
+        internal static class HttpHandlerDefaults
+        {
+            public static int MaxConnectionsPerServer { get; } = GetMaxConnectionsPerServer();
+
+            private static int GetMaxConnectionsPerServer()
+            {
+                int value = RuntimeSettingParser.QueryRuntimeSettingInt32(
+                    "System.Net.SocketsHttpHandler.MaxConnectionsPerServer",
```
Comment (ManickaP) [inline src/libraries/System.Net.Http/src/System/Net/Http/GlobalHttpSettings.cs:88]: If this is named `SocketsHttpHandler` it shouldn't affect `WinHttpHandler`. Or it should be named differently, depending on what's the goal of the setting.

Comment (rzikm) [reply src/libraries/System.Net.Http/src/System/Net/Http/GlobalHttpSettings.cs:88]: Scratch that, I can't reasonably propagate the value to WinHttpHandler (it is built for .NET Framework as well and there are unfortunate differences which prevent including necessary .cs files).

Comment (ManickaP) [reply src/libraries/System.Net.Http/src/System/Net/Http/GlobalHttpSettings.cs:88]: Is there a reason this is not part of `internal static class SocketsHttpHandler` on line 20? Especially now that you decided to scope it purely to sockets handler.

Comment (rzikm) [reply src/libraries/System.Net.Http/src/System/Net/Http/GlobalHttpSettings.cs:85]: No, I forgot to move it. Should be done now

**Review by ManickaP — APPROVED** [review]
Small nits, otherwise LGTM, thanks!

**Pull request merged**